OpenRA.Mods.Common/Traits/Harvester.cs

```diff
@@ -250,28 +248,27 @@ public virtual bool TickUnload(Actor self, Actor proc)
 			if (--currentUnloadTicks > 0)
 				return false;
 
-			if (contents.Keys.Count > 0)
+			var acceptResources = proc.Trait<IAcceptResources>();
+			foreach (var sr in storesResources)
 			{
-				var acceptResources = proc.Trait<IAcceptResources>();
-				foreach (var c in contents)
+				if (sr.ContentsSum == 0)
+					continue;
+
+				foreach (var c in sr.Contents)
 				{
-					var resourceType = c.Key;
 					var count = Math.Min(c.Value, Info.BaleUnloadAmount);
-					var accepted = acceptResources.AcceptResources(resourceType, count);
+					var accepted = acceptResources.AcceptResources(c.Key, count);
 					if (accepted == 0)
 						continue;
 
-					contents[resourceType] -= accepted;
-					if (contents[resourceType] <= 0)
-						contents.Remove(resourceType);
-
+					sr.RemoveResource(c.Key, accepted);
 					currentUnloadTicks = Info.BaleUnloadDelay;
 					UpdateCondition(self);
 					return false;
 				}
 			}
 
-			return contents.Count == 0;
+			return IsEmpty;
```

Comment:
Seems strange/buggy, already before? In the above loop you try to find resources to unload. If not you exit or don't enter the loop, and then here again you check the total resources to again determine if empty?

Shouldn't this just always be `true`?